include/dali/core/mm/pool_resource.h

```diff
@@ -59,11 +86,159 @@ constexpr pool_options default_device_pool_opts() noexcept {
   return { (static_cast<size_t>(1) << 32), (1 << 20), 2.0f, true, true };
 }
 
+template <memory_kind kind>
+constexpr sync_scope default_sync_scope() {
+  return kind == memory_kind::device ? sync_scope::device
+                                     : kind == memory_kind::host ? sync_scope::none
+                                                                 : sync_scope::system;
+}
+
+template <memory_kind kind>
+constexpr pool_options default_pool_opts() noexcept {
+  if (kind == memory_kind::host) {
+    return default_host_pool_opts();
+  } else {
+    auto opt = default_device_pool_opts();
+    opt.sync = default_sync_scope<kind>();
+    opt.enable_deferred_deallocation = true;
+    return opt;
+  }
+}
+
+struct dealloc_params {
+  int sync_device = -1;  // -1 == current device
+  void *ptr = nullptr;
+  size_t bytes = 0, alignment = 0;
+};
+
+namespace detail {
+template <typename Actual>
+class deferred_deallocator {
+ public:
+  using dealloc_queue = SmallVector<dealloc_params, 32>;
+
+  ~deferred_deallocator() {
+    if (worker_.joinable()) {
+      stop();
+      worker_.join();
+    }
+  }
+
+  void deferred_deallocate(void *ptr,
+                           size_t bytes,
+                           size_t alignment = alignof(std::max_align_t),
+                           int device_id = -1) {
+    if (!ptr || !bytes)
+      return;  // nothing to do
+    if (device_id < 0) {
+      CUDA_CALL(cudaGetDevice(&device_id));
+    }
+
+    {
+      std::lock_guard<std::mutex> g(mtx_);
+      deallocs_[queue_idx_].push_back({device_id, ptr, bytes, alignment});
+
+      if (!started_)
```

Comment:
Maybe `start_worker` should check this condition?

Reply:
I found it more human readable this way "if not already started then start" but I can move it if you insist.

Reply:
Just asking. No strong opinion.